ext/random/random.c

```diff
@@ -0,0 +1,871 @@
+		zend_call_known_instance_method_with_0_params(
+			php_random_ce_Random_NumberGenerator_XorShift128Plus->constructor,
+			rng,
+			NULL
+		);
+		/* No needed self-refcount */
+		rng->gc.refcount--;
+	}
+
+	ZVAL_OBJ(&zrng, rng);
+
+	pname = zend_string_init("rng", sizeof("rng") - 1, 0);
+	zend_std_write_property(&random->std, pname, &zrng, NULL);
+	zend_string_release(pname);
+
+	/* Internal class always php_random_rng */
+	if (rng->ce->type == ZEND_INTERNAL_CLASS) {
+		random->rng = php_random_rng_from_obj(rng);
+	}
+}
+
+PHP_METHOD(Random, getNumberGenerator)
+{
+	php_random *random = Z_RANDOM_P(ZEND_THIS);
+	zval *zrng;
+
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	if (random->rng) {
+		RETURN_OBJ_COPY(&random->rng->std);
+	}
+
+	zrng = zend_read_property(random->std.ce, &random->std, "rng", sizeof("rng") - 1, 0, NULL);
+	RETURN_OBJ(Z_OBJ_P(zrng));
+}
+
+PHP_METHOD(Random, nextInt)
+{
+	php_random *random = Z_RANDOM_P(ZEND_THIS);
+
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	RETURN_LONG((zend_long) php_random_next(random, 1));
+}
+
+PHP_METHOD(Random, getInt)
+{
+	php_random *random = Z_RANDOM_P(ZEND_THIS);
+	zend_long min, max;
+
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_LONG(min)
+		Z_PARAM_LONG(max)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (UNEXPECTED(max < min)) {
+		zend_argument_value_error(2, "must be greater than or equal to argument #1 ($min)");
+		RETURN_THROWS();
+	}
+
+	RETURN_LONG(php_random_range(random, min, max));
+}
+
+PHP_METHOD(Random, getBytes)
+{
+	php_random *random = Z_RANDOM_P(ZEND_THIS);
+	zend_string *ret;
+	zend_long size;
+	uint64_t buf;
+	uint8_t *bytes;
+	size_t generated_bytes = 0;
+	int i;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_LONG(size)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (size < 1) {
+		zend_argument_value_error(1, "must be greater than 0");
+		RETURN_THROWS();
+	}
+
+	ret = zend_string_alloc(size, 0);
+
+	while (generated_bytes <= size) {
+		buf = php_random_next(random, 0);
+		if (random->rng && random->rng->algo->generate_size == sizeof(uint32_t)) {
+			buf = (buf << 32) | php_random_next(random, 0);
+		}
+		bytes = (uint8_t *) &buf;
+		for (i = 0; i < (sizeof(uint64_t) / sizeof(uint8_t)); i ++) {
+			ZSTR_VAL(ret)[generated_bytes + i] = bytes[i];
+
+			if ((generated_bytes + i) >= size) {
+				ZSTR_VAL(ret)[size] = '\0';
+				RETURN_STR(ret);
+			}
+		}
+		generated_bytes += (sizeof(uint64_t) / sizeof(uint8_t));
+	}
+}
+
+PHP_METHOD(Random, shuffleArray)
+{
+	php_random *random =  Z_RANDOM_P(ZEND_THIS);;
+	zval *array;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY(array)
+	ZEND_PARSE_PARAMETERS_END();
+
+	ZVAL_DUP(return_value, array);
+	php_random_array_data_shuffle(random, return_value);
+}
+
+PHP_METHOD(Random, shuffleString)
+{
+	php_random *random = Z_RANDOM_P(ZEND_THIS);
+	zend_string *string;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_STR(string)
+	ZEND_PARSE_PARAMETERS_END();
+
+	RETVAL_STRINGL(ZSTR_VAL(string), ZSTR_LEN(string));
+	if (Z_STRLEN_P(return_value) > 1) {
```

Comment:
If you want to optimize this case, it would make more sense to handle it first using:
```
if (ZSTR_LEN(string) < 2) {
    RETURN_STR_COPY(string);
}
```

Reply:
Fixed Thanks